scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala

```diff
@@ -12,327 +13,194 @@ import scala.compat.Platform.EOL
+  allowMultilinePrograms: Boolean,
 
-  // Are XML literals supported by this dialect?
-  // We plan to deprecate XML literal syntax, and some dialects
-  // might go ahead and drop support completely.
-  def allowXmlLiterals: Boolean
+  // Are `|` (union types) supported by this dialect?
+  allowOrTypes: Boolean,
 
-  // Are inline vals and defs supported by this dialect?
-  def allowInline: Boolean
+  // Are unquotes ($x) and splices (..$xs, ...$xss) allowed?
+  // If yes, they will be parsed as patterns.
+  allowPatUnquotes: Boolean,
+
+  // Are naked underscores allowed after $ in pattern interpolators, i.e. is `case q"$_ + $_" =>` legal or not?
+  allowSpliceUnderscores: Boolean,
+
+  // Are unquotes ($x) and splices (..$xs, ...$xss) allowed?
+  // If yes, they will be parsed as terms.
+  allowTermUnquotes: Boolean,
 
   // Are terms on the top level supported by this dialect?
   // Necessary to support popular script-like DSLs.
-  def allowToplevelTerms: Boolean
+  allowToplevelTerms: Boolean,
 
-  // Are `|` (union types) supported by this dialect?
-  def allowOrTypes: Boolean
+  // Are trailing commas allowed? SIP-27.
+  allowTrailingCommas: Boolean,
 
-  // What kind of separator is necessary to split top-level statements?
-  // Normally none is required, but scripts may have their own rules.
-  def toplevelSeparator: String
+  // Are trait allowed to have parameters?
+  // They are in Dotty, but not in Scala 2.12 or older.
+  allowTraitParameters: Boolean,
 
   // Are view bounds supported by this dialect?
   // Removed in Dotty.
-  def allowViewBounds: Boolean
-
-  // Are `&` intersection types supported by this dialect?
-  def allowAndTypes: Boolean
+  allowViewBounds: Boolean,
 
   // Are `with` intersection types supported by this dialect?
-  def allowWithTypes: Boolean = !allowAndTypes
+  allowWithTypes: Boolean,
 
-  // Are trait allowed to have parameters?
-  // They are in Dotty, but not in Scala 2.12 or older.
-  def allowTraitParameters: Boolean
-
-  // Are literal types allowed, i.e. is `val a : 42 = 42` legal or not?
-  def allowLiteralTypes: Boolean
+  // Are XML literals supported by this dialect?
+  // We plan to deprecate XML literal syntax, and some dialects
+  // might go ahead and drop support completely.
+  allowXmlLiterals: Boolean,
 
-  // Are trailing commas allowed? SIP-27.
-  def allowTrailingComma: Boolean
+  // What kind of separator is necessary to split top-level statements?
+  // Normally none is required, but scripts may have their own rules.
+  toplevelSeparator: String
+) {
+  // Are unquotes ($x) and splices (..$xs, ...$xss) allowed?
+  def allowUnquotes: Boolean = allowTermUnquotes || allowPatUnquotes
+
+  // Dialects have reference equality semantics,
+  // because sometimes dialects representing distinct Scala versions
+  // can be structurally equal to each other.
+  override def canEqual(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
```

Comment:
> If name is automatically populated by a macro, then it won't be part of the case class signature,

Not sure I follow.

> Wouldn't custom equality and custom toString still be a thing?

Sure, custom equality and toString are fine by me. My suggestion was only to make Dialect as easy as possible to understand by making it a plain `@data` class, but I don't really mind the customization.

> As for using JVM reflection, I don't mind changing Dialect.standard to be populated manually.

We can keep reflection for now, but add a comment for our future future self that this can easily be replaced with a manual list to support Scala.js.